### PR TITLE
Jeff/step 14

### PR DESCRIPTION
**Pull request by jeff-yoo-rakuma**
# Overview
This PR adds pagination params, `offset` and `limit`, to GET endpoint. 
It will work exactly like SQL's offset and limit.

# Changes
## Backend
* Add `offset` and `limit` params to GET endpoint
## Frontend
* Add infinite scrolling to utilize pagination function

# Tests
```
root@d992cc5219f1:/app# rspec
...............................

Finished in 2.53 seconds (files took 7.39 seconds to load)
31 examples, 0 failures
```

### CODE REVIEW

**Comment by shunshunsuke** [comment]
@jeff-yoo-rakuma 
I ran the following command, but there was no pagination. 😢 
```
docker-compose build
docker-compose up -d
```

Is there anything else I need to do?

**Comment by shunshunsuke** [comment]
https://user-images.githubusercontent.com/61672778/148033571-ad477d7d-a02f-465a-94ec-cb945e7669cb.mp4

**Comment by jeff-yoo-rakuma** [comment]
> @jeff-yoo-rakuma I ran the following command, but there was no pagenation. 😢
> 
> ```
> docker-compose build
> docker-compose up -d
> ```
> 
> Is there anything else I need to do?

@shunshunsuke 
The pagination is implemented with infinite scrolling in my case 😅
As you scroll down, you can see more todos are loaded. I thought it suits better in my grid design.
You can see the logs like these if you have docker compose running on foreground
```
task-service    | Started GET "/tasks?sort=created_at:desc&offset=0&limit=20" for 172.21.0.1 at 2022-01-04 17:57:27 +0900
task-service    | Processing by TasksController#index as HTML
task-service    |   Parameters: {"sort"=>"created_at:desc", "offset"=>"0", "limit"=>"20"}
task-service    |   Task Load (0.7ms)  SELECT  `tasks`.* FROM `tasks` ORDER BY `tasks`.`created_at` DESC LIMIT 20 OFFSET 0
task-service    | Completed 200 OK in 7ms (Views: 5.6ms | ActiveRecord: 0.7ms)
task-service    |
task-service    |
task-service    | Started GET "/tasks?sort=created_at:desc&offset=20&limit=20" for 172.21.0.1 at 2022-01-04 17:57:28 +0900
task-service    | Processing by TasksController#index as HTML
task-service    |   Parameters: {"sort"=>"created_at:desc", "offset"=>"20", "limit"=>"20"}
task-service    |   Task Load (0.6ms)  SELECT  `tasks`.* FROM `tasks` ORDER BY `tasks`.`created_at` DESC LIMIT 20 OFFSET 20
task-service    | Completed 200 OK in 6ms (Views: 4.6ms | ActiveRecord: 0.6ms)
task-service    |
task-service    |
task-service    | Started GET "/tasks?sort=created_at:desc&offset=40&limit=20" for 172.21.0.1 at 2022-01-04 17:57:29 +0900
task-service    | Processing by TasksController#index as HTML
task-service    |   Parameters: {"sort"=>"created_at:desc", "offset"=>"40", "limit"=>"20"}
task-service    |   Task Load (0.6ms)  SELECT  `tasks`.* FROM `tasks` ORDER BY `tasks`.`created_at` DESC LIMIT 20 OFFSET 40
task-service    | Completed 200 OK in 4ms (Views: 3.5ms | ActiveRecord: 0.6ms)
```

File: todos-app/backend/task_service/app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -68,4 +69,28 @@ def status_param
...
   end
+
+  def pagination_params
+    if params[:offset].nil?
+      @offset = 0
+    else
+      begin
+        @offset = Integer(params[:offset])
+      rescue ArgumentError
+        raise Exceptions::InvalidPaginationParams, "Given offset #{params[:offset]} is not integer"
```
Comment (shunshunsuke) [inline todos-app/backend/task_service/app/controllers/tasks_controller.rb:80]: 👍

**Comment by shunshunsuke** [comment]
Oh, I see!
I had assumed that pagination was something like the following.
image

**Review by shunshunsuke — APPROVED** [review]
LGTM!
LGTM

**Comment by jeff-yoo-rakuma** [comment]
> Oh, I see! I had assumed that pagination was something like the following. image

That's true in the traditional sense! The backend implementation will be the same either way so I just decided to have frontend do infinite scrolling!

**Pull request merged**